addons/mass_mailing/models/mailing.py

```diff
@@ -260,15 +260,34 @@ def _compute_statistics(self):
             self.browse(row.pop('mailing_id')).update(row)
 
     def _compute_next_departure(self):
-        cron_next_call = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo().nextcall
+        cron_job = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo()
         str2dt = fields.Datetime.from_string
-        cron_time = str2dt(cron_next_call)
+        cron_time = str2dt(cron_job.nextcall)
+        cron_id = cron_job.id
+
+        # A cron.trigger is created when mailing is put "in queue"
+        # The cron workers usually clean up old cron.triggers after finishing their
+        # jobs, i.e older than now(). Thus only most recent than now is relevant
+        now_datetime_buffer = fields.Datetime.now() + relativedelta(minutes=1)
+        cron_triggers = self.env['ir.cron.trigger'].sudo().search([
+            '&',
+            ('cron_id', '=', cron_id),
+            ('call_at', '<', now_datetime_buffer)],
+            order='call_at desc', limit=1)
+        most_recent_trigger_datetime = cron_triggers.call_at
         for mass_mailing in self:
             if mass_mailing.schedule_date:
-                schedule_date = str2dt(mass_mailing.schedule_date)
-                mass_mailing.next_departure = max(schedule_date, cron_time)
+                mass_mailing.next_departure = max(mass_mailing.schedule_date, fields.datetime.now())
```

Comment:
Joining in on this, for this use case we just assume that there will be a cron trigger created at schedule time and we do not check its existence?

Seems a bit simple, but I'm guessing all the flows cover that? (Putting in queue / editing scheduled date / ...) ?
Makes me wonder why we are checking that it exists when sending immediately though.

Trying to fetch all cron triggers would be more complex, as you could have "self" containing multiple mailings with different schedule date.

Maybe I missed something.

Reply:
Thx for the input.

As you guessed I am working with the implicit nature for the flow here
```   
    def action_launch(self):
        self.write({'schedule_type': 'now'})
        return self.action_put_in_queue()

    def action_schedule(self):
        self.ensure_one()
        if self.schedule_date:
            return self.action_put_in_queue()
        else:
            action = self.env["ir.actions.actions"]._for_xml_id("mass_mailing.mailing_mailing_schedule_date_action")
            action['context'] = dict(self.env.context, default_mass_mailing_id=self.id)
            return action

    def action_put_in_queue(self):
        self.write({'state': 'in_queue'})
        cron = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue')
        cron._trigger(
            schedule_date or fields.Datetime.now()
            for schedule_date in self.mapped('schedule_date')
        )
```

The thing is that `mass_mailing.schedule_date` only gets set when using the schedule wizard, i.e. stays False when using "Send" button (which goes directly to `.action_launch`) .

Now for this code line in question, we could fetch the trigger, but the issue is that if the user schedules a date way in the future ( after now or nextcall), we would need to add an extra `search()` matching the datetime of the `schedule_date` field (since the cron.trigger model has no other metadata than cronjob id and call_at).

This extra calculation step might not be necessary as ` action_put_in_queue` will generate a trigger for that datetime if `schedule_date` is set. So we can reasonably well expect that unless we have an uncaught error, the trigger is created.

Reply:
Then the follow-up question was, why do we assume it exists at the scheduled_date and not assume it exists when sending immediately?

Is that "extra minute" important to say it will be sent at 09:00 AM or at 09:01 AM?

I feel like if we assume the flow will create the triggers, then we should always assume it will.
And if we don't, then we check (all) created triggers.

Currently it's a bit of a mix of both.

Reply:
You are right. The current version was the natural progression of the input from reth's comments.

But now that you point it out we could simplify the definition of `_compute_next_departure` and trust the flow, as any small time discrepancy would not have any real world impact.

What I can do is export the cron.trigger checks into the unit test, and simply assert if the relevant cron.trigger was indeed created. So if the cron.trigger flow ever changes in the future the tests should catch it.

Should I do that?

Reply:
I think that's fair, but maybe keep your current version somewhere in case other reviewers don't agree :slightly_smiling_face:

Reply:
@tde-banana-odoo Understandable!
So after awa's input lead me to an epiphany, I overhauled the whole thing. 
In summary:
- We can keep `_compute_next_departure` very simple, since the methods `action_launch` and
`action_schedule` should take care of the cron.trigger creation. Thus either we are gonna use the schedule_date or now() depending on if it is a scheduled mass mailing campaign or a "Send now" (i.e. no schedule_date set). Noo need to fetch the cron job id or cron time etc.
- I migrated all the check to see if the cron.triggers are created as expected into the unit tests. A long as those don't fail, there should be not reason to believe that our simplified version of `_compute_next_departure` should behave unexpectedly

Now if you don't like this approach I could revert the changes, but in my humble opinion this seems like a cleaner solution.

Reply:
In short:

When scheduling / posting, we currently create triggers (in action_put_in_queue).
So for the next departure we assume that triggers will be created and we display "now" if it's an immediate send and the schedule_date if it's scheduled.
And we add tests to ensure that triggers are correctly created.

(Correct me if I'm wrong here @jorv-odoo ).

If we don't want it to be that "simple", then we have to fetch all created triggers depending on shceduled_date and runtime of the cron, and it becomes much more complex.

Reply:
Just to be sure: if triggers disappears, next sending possibility will anyway be the cron running itself, like once / day ?

Reply:
The back-end side yes and per default that specific cron job runs once a day.
As long as the stats is in queue and schedule date < now, that mailing campaign should run
```
        mass_mailings = self.search([('state', 'in', ('in_queue', 'sending')), '|', ('schedule_date', '<', fields.Datetime.now()), ('schedule_date', '=', False)])
```

So if someone or something removes the trigger, the displayed next_departure would be slightly off time-wise depending on the cron job nextcall value.